effectInFailure: plugin-only TS2589 from a bare `import "@cloudflare/workers-types"`

### What version of Effect is running?

- `@effect/tsgo`: `0.5.1` (also confirmed on `0.4.0`)
- `@typescript/native-preview`: `7.0.0-dev.20260425.1`
- macOS `26.3.1`, `darwin-arm64`

### What steps can reproduce the bug?

Self-contained repro: https://github.com/danieljvdm/effect-tsgo-cf-ts2589-repro

```sh
git clone https://github.com/danieljvdm/effect-tsgo-cf-ts2589-repro.git
cd effect-tsgo-cf-ts2589-repro
bun install
bun run verify
```

The user program is one line:

```ts
// src/repro.ts
import "@cloudflare/workers-types";
```

Three tsconfigs toggle just the plugin / one rule:

| Config | Plugin? | `effectInFailure` | Result |
| --- | --- | --- | --- |
| `tsconfig.json` | enabled | default | **TS2589** |
| `tsconfig.no-plugin.json` | removed | — | clean |
| `tsconfig.rule-off.json` | enabled | `"off"` | clean |

`--extendedDiagnostics` shows total instantiations jump 12× when the rule fires (2,090 → 25,040), which is what trips tsgo's depth budget.

### What is the expected behavior?

`tsgo --noEmit -p tsconfig.json` should typecheck cleanly. The user program contains no Effect code; nothing in it should be reachable for the `effectInFailure` rule.

### What do you see instead?

```
error TS2589: Type instantiation is excessively deep and possibly infinite.
```

Sub-second, no source location attached.

### Additional information

This three-way matrix matches the assertions in [`internal/effecttest/effect_in_failure_ts2589_test.go`](https://github.com/Effect-TS/tsgo/blob/8869556/internal/effecttest/effect_in_failure_ts2589_test.go) added in [#167](https://github.com/Effect-TS/tsgo/pull/167) / [`8869556`](https://github.com/Effect-TS/tsgo/commit/8869556) — but the trigger here is a real-world bare import of a large ambient `.d.ts` module, not a hand-built deeply nested generic. The `0.5.0` typeparser refactor reduces overhead for the test's failure shape but does not address this codepath. Reproduces identically on `0.4.0`, so it is not a `0.5.x` regression.

Same TS2589 (with the same instantiation explosion) reproduces on any of:

- `import "@cloudflare/workers-types";`
- `import "@cloudflare/workers-types/experimental";`
- `import "wrangler";`
- `import "miniflare";`
- `import "@cloudflare/vite-plugin";`

…each of which transitively reaches `@cloudflare/workers-types`. It does not reproduce on `import "vite";`, so the trigger is the size + shape of the `@cloudflare/workers-types` type graph, not module loading per se.

Suspected cause: `internal/rules/effect_in_failure.go` does a post-order AST walk and calls `GetTypeAtLocation` + `StrictEffectType` on every node. On the import binding, that lazily expands the entire imported namespace's type graph through the underlying tsgo checker; `StrictEffectType` correctly returns `nil` for the non-Effect types, but only after the checker work is already done. A cheap-bail before invoking the checker on AST node kinds that cannot resolve to an Effect (e.g. `ImportClause`, `ImportSpecifier`, raw string literals) would likely fix this.

Workaround I'm using in the meantime: per-program `diagnosticSeverity: { "effectInFailure": "off" }` in the affected `tsconfig.json`'s plugin entry.

--- a/.changeset/calm-cloudflare-types.md
+++ b/.changeset/calm-cloudflare-types.md
@@ -1,0 +1,5 @@
+---
+"@effect/tsgo": patch
+---
+
+Skip Effect diagnostics for source files resolved from external libraries.

--- a/_patches/002-checker-checker.patch
+++ b/_patches/002-checker-checker.patch
@@ -43,6 +43,14 @@ index 2e1622418..66059cab4 100644
  // CheckMode
  
  type CheckMode uint32
+@@ -559,6 +593,7 @@ type Program interface {
+ 	GetImportHelpersImportSpecifier(path tspath.Path) *ast.Node
+ 	SourceFileMayBeEmitted(sourceFile *ast.SourceFile, forceDtsEmit bool) bool
+ 	IsSourceFileDefaultLibrary(path tspath.Path) bool
++	IsSourceFileFromExternalLibrary(file *ast.SourceFile) bool
+ 	GetProjectReferenceFromOutputDts(path tspath.Path) *tsoptions.SourceOutputAndProjectReference
+ 	GetRedirectForResolution(file ast.HasFileName) *tsoptions.ParsedCommandLine
+ 	CommonSourceDirectory() string
 @@ -2145,6 +2179,8 @@ func (c *Checker) checkSourceFile(ctx context.Context, sourceFile *ast.SourceFil
  	links := c.sourceFileLinks.Get(sourceFile)
  	if !links.typeChecked {

--- a/_patches/018-ls-autoimport-stylepolicy.patch
+++ b/_patches/018-ls-autoimport-stylepolicy.patch
@@ -126,3 +126,18 @@ index b2107443f..bf5161047 100644
  }
  
  // getAddAsTypeOnly determines if an import should be type-only based on usage context
+diff --git a/internal/ls/autoimport/aliasresolver.go b/internal/ls/autoimport/aliasresolver.go
+--- a/internal/ls/autoimport/aliasresolver.go
++++ b/internal/ls/autoimport/aliasresolver.go
+@@ -220,6 +220,11 @@ func (r *aliasResolver) IsSourceFileDefaultLibrary(path tspath.Path) bool {
+ 	panic("unimplemented")
+ }
+ 
++// IsSourceFileFromExternalLibrary implements checker.Program.
++func (r *aliasResolver) IsSourceFileFromExternalLibrary(file *ast.SourceFile) bool {
++	return false
++}
++
+ // IsSourceFromProjectReference implements checker.Program.
+ func (r *aliasResolver) IsSourceFromProjectReference(path tspath.Path) bool {
+ 	panic("unimplemented")

--- a/etscheckerhooks/external_library_test.go
+++ b/etscheckerhooks/external_library_test.go
@@ -1,0 +1,98 @@
+package etscheckerhooks_test
+
+import (
+	"context"
+	"strings"
+	"testing"
+	"testing/fstest"
+
+	_ "github.com/effect-ts/tsgo/etscheckerhooks"
+	"github.com/effect-ts/tsgo/etscore"
+	"github.com/microsoft/typescript-go/shim/ast"
+	"github.com/microsoft/typescript-go/shim/bundled"
+	"github.com/microsoft/typescript-go/shim/compiler"
+	"github.com/microsoft/typescript-go/shim/core"
+	"github.com/microsoft/typescript-go/shim/tsoptions"
+	"github.com/microsoft/typescript-go/shim/vfs/vfstest"
+)
+
+func TestEffectDiagnosticsSkipExternalLibrarySourceFiles(t *testing.T) {
+	t.Parallel()
+
+	testfs := map[string]any{
+		"/src/main.ts": &fstest.MapFile{
+			Data: []byte(`import { ExternalProblem } from "problem"
+
+export class LocalProblem extends Error {}
+export const value = ExternalProblem
+`),
+		},
+		"/node_modules/problem/package.json": &fstest.MapFile{
+			Data: []byte(`{"name":"problem","version":"1.0.0","types":"index.ts"}`),
+		},
+		"/node_modules/problem/index.ts": &fstest.MapFile{
+			Data: []byte(`export class ExternalProblem extends Error {}`),
+		},
+	}
+
+	fs := bundled.WrapFS(vfstest.FromMap(testfs, true))
+	options := &core.CompilerOptions{
+		NewLine:           core.NewLineKindLF,
+		NoErrorTruncation: core.TSTrue,
+		Target:            core.ScriptTargetESNext,
+		Module:            core.ModuleKindCommonJS,
+		ModuleResolution:  core.ModuleResolutionKindNode10,
+		Effect: &etscore.EffectPluginOptions{
+			Diagnostics: true,
+			DiagnosticSeverity: map[string]etscore.Severity{
+				"extendsNativeError": etscore.SeverityError,
+			},
+		},
+	}
+	program := compiler.NewProgram(compiler.ProgramOptions{
+		Config: &tsoptions.ParsedCommandLine{
+			ParsedConfig: &core.ParsedOptions{
+				CompilerOptions: options,
+				FileNames:       []string{"/src/main.ts"},
+			},
+		},
+		Host:           compiler.NewCompilerHost("/", fs, bundled.LibPath(), nil, nil),
+		SingleThreaded: core.TSTrue,
+	})
+
+	dependencySource := program.GetSourceFile("/node_modules/problem/index.ts")
+	if dependencySource == nil {
+		t.Fatal("expected dependency source file to be resolved")
+	}
+	if !program.IsSourceFileFromExternalLibrary(dependencySource) {
+		t.Fatal("expected dependency source file to be marked as external library")
+	}
+
+	diagnostics := program.GetSemanticDiagnostics(context.Background(), nil)
+	localEffectDiagnostics := 0
+	externalEffectDiagnostics := 0
+	for _, diag := range diagnostics {
+		if !isEffectDiagnostic(diag) || diag.File() == nil {
+			continue
+		}
+		fileName := diag.File().FileName()
+		switch {
+		case strings.Contains(fileName, "/node_modules/"):
+			externalEffectDiagnostics++
+		case fileName == "/src/main.ts":
+			localEffectDiagnostics++
+		}
+	}
+
+	if externalEffectDiagnostics != 0 {
+		t.Fatalf("expected no Effect diagnostics from external libraries, got %d", externalEffectDiagnostics)
+	}
+	if localEffectDiagnostics == 0 {
+		t.Fatal("expected Effect diagnostics for project source file")
+	}
+}
+
+func isEffectDiagnostic(diag *ast.Diagnostic) bool {
+	code := diag.Code()
+	return code >= 377000 && code < 378000
+}

--- a/etscheckerhooks/init.go
+++ b/etscheckerhooks/init.go
@@ -45,7 +45,9 @@ type RuleDiagnostic struct {
 // afterCheckSourceFile is called after type checking each source file.
 // It runs Effect diagnostics if the plugin is enabled.
 func afterCheckSourceFile(ctx context.Context, program checker.Program, c *checker.Checker, sf *ast.SourceFile) {
-	tp := typeparser.NewTypeParser(program, c)
+	if sf.IsDeclarationFile || program.IsSourceFileFromExternalLibrary(sf) {
+		return
+	}
 
 	// Get Effect config from program options (parsed during config loading)
 	effectConfig := getEffectConfig(program)
@@ -71,10 +73,7 @@ func afterCheckSourceFile(ctx context.Context, program checker.Program, c *check
 		return
 	}
 
-	// Skip declaration files
-	if sf.IsDeclarationFile {
-		return
-	}
+	tp := typeparser.NewTypeParser(program, c)
 
 	// Collect directives from source file for suppression support
 	sourceText := sf.Text()

